server/src/main/java/org/elasticsearch/index/codec/vectors/ES813FlatVectorFormat.java

```diff
@@ -0,0 +1,151 @@
+    }
+
+    @Override
+    public KnnVectorsReader fieldsReader(SegmentReadState state) throws IOException {
+        return new ES813FlatVectorReader(format.fieldsReader(state));
+    }
+
+    public static class ES813FlatVectorWriter extends KnnVectorsWriter {
+
+        private final FlatVectorsWriter writer;
+
+        public ES813FlatVectorWriter(FlatVectorsWriter writer) {
+            super();
+            this.writer = writer;
+        }
+
+        @Override
+        public KnnFieldVectorsWriter<?> addField(FieldInfo fieldInfo) throws IOException {
+            return writer.addField(fieldInfo, null);
+        }
+
+        @Override
+        public void flush(int maxDoc, Sorter.DocMap sortMap) throws IOException {
+            writer.flush(maxDoc, sortMap);
+        }
+
+        @Override
+        public void finish() throws IOException {
+            writer.finish();
+        }
+
+        @Override
+        public void close() throws IOException {
+            writer.close();
+        }
+
+        @Override
+        public long ramBytesUsed() {
+            return writer.ramBytesUsed();
+        }
+
+        @Override
+        public void mergeOneField(FieldInfo fieldInfo, MergeState mergeState) throws IOException {
+            writer.mergeOneField(fieldInfo, mergeState);
+        }
+    }
+
+    public static class ES813FlatVectorReader extends KnnVectorsReader {
+
+        private final FlatVectorsReader reader;
+
+        public ES813FlatVectorReader(FlatVectorsReader reader) {
+            super();
+            this.reader = reader;
+        }
+
+        @Override
+        public void checkIntegrity() throws IOException {
+            reader.checkIntegrity();
+        }
+
+        @Override
+        public FloatVectorValues getFloatVectorValues(String field) throws IOException {
+            return reader.getFloatVectorValues(field);
+        }
+
+        @Override
+        public ByteVectorValues getByteVectorValues(String field) throws IOException {
+            return reader.getByteVectorValues(field);
+        }
+
+        @Override
+        public void search(String field, float[] target, KnnCollector knnCollector, Bits acceptDocs) throws IOException {
+            collectAllMatchingDocs(knnCollector, acceptDocs, reader.getRandomVectorScorer(field, target));
+        }
+
+        private void collectAllMatchingDocs(KnnCollector knnCollector, Bits acceptDocs, RandomVectorScorer scorer) throws IOException {
+            OrdinalTranslatedKnnCollector collector = new OrdinalTranslatedKnnCollector(knnCollector, scorer::ordToDoc);
+            Bits acceptedOrds = scorer.getAcceptOrds(acceptDocs);
+            for (int i = 0; i < scorer.maxOrd(); i++) {
+                if (acceptedOrds == null || acceptedOrds.get(i)) {
+                    collector.collect(i, scorer.score(i));
+                    collector.incVisitedCount(1);
+                }
+            }
+        }
```

Comment:
Should it early exit if the number of vectors is greater than the limit on the knn collector? This way we would not double-collect `visitLimit()` docs when filtering is enabled? Or is it illegal?

Reply:
> I'm also curious to better understand if there's some inefficiency around filtered search with the flat format, if I understand correctly we would often first score the first knnCollector.visitLimit() doc to then score all docs that match the filter. It would be nice if we could somehow skip the unfiltered linear scan all the time?

Yeah, this is frustrating :(. Its because we consider "early terminated" to be `>=` the limit, but we will hit the limit every time as the limit == number of docs in the filter.

I think this will be possible by adding a new KnnQuery kind that is returned when the query is over a flat-codec.

Reply:
I wonder if this usage should be prohibited for the `flat` format (like throwing an exception when a flat field is used in a knn query). It feels wrong to use the `knn` section/query if it's "just" to score documents. For a flat use case, users should rely on the function score query or a sort criteria that is inlined in the main query. The complexity of the `knn` section with its own filtering can be reserved for approximate use cases that require a different strategy?

Reply:
@jimczi yeah, I was also thinking something similar. The tricky part is that we want to use the `RandomVectorScorer` here provided by the internal `FlatVectorFormat`. Iterating and scoring the vectors as normal would bypass any benefit from int8 as we would end up using and loading the float32 vectors.

Another option would be to have an interface in ES (`VectorScorer...something`). This coupled with a custom kNN query implementation that uses the `RandomVectorScorer` provided by the flat codecs. 

This requires extracting the `knnVectorReader` from the leafContext though.

I do not see a clean interface all the way down to Lucene without doing stuff like:
```
if (context.reader() instanceof CodecReader codecReader) {
...get knnReader...
...check that it provides RandomVectorScorer....
}
```

Reply:
@jimczi I do not think however, that a user using a `knn` query against multiple dense_vector indices where one is `flat` and the other is `hnsw` should result in an error. 

We should just do the correct thing internally.

Reply:
@benwtrent Returning a different query based on whether the format is flat or not sounds like a good approach. :+1:

Reply:
That would work for the `knn` section/query @jpountz but there are other contexts we need to handle. 
I think this PR is doing the right thing, it allows to use the `flat` format in all contexts (scripts and top N `knn` section/query). In a follow up we can think of how we unleash the auto-int8 quantisation similarity in scripts or in a new brute force query. The point being that the `knn` section/query is just one way to use the `flat` format and probably not the preferred one.

Reply:
> return a query that works directly on `VectorValues` when the format is flat, akin to `SortedNumericDocValuesRangeQuery`.

When vectors are stored in a lossy format (like `int8` or `pq`), it is useful to be able to not only search, but rescore. 

Right now, I was thinking `script_score` is a rescoring phase (e.g. the user asks for a cosine/dotProduct function directly, regardless of the index's configured similarity). This would even be useful for `int8` and especially for other quantizations like `pq` & `int4`.

Using `VectorValues` directly will not work for any quantized vector data as additional information is required:

 - The appropriate scoring & correction
 - Quantizing the actual query vector, which will be different per segment.

This is why we have the `RandomVectorScorer` interface, it encapsulates that.

As for how we expose quantized/lossy comparisons as function scripts, I honestly don't know.

Maybe there is a new "vectorSimilarity" function that requires the field to be `indexed` and uses the stored vector similarity & quantized (or not) vector values.

Reply:
@jpountz I don't think we can do a specialized query for this until the Lucene interfaces are better for flat codecs. I ended up having to do something like this because the LeafReader just doesn't have a good interface for this:

```
    private RandomVectorScorer getScorer(LeafReaderContext context) throws IOException {
        assert floatVector != null || byteVector != null;
        LeafReader reader = FilterLeafReader.unwrap(context.reader());
        if (reader instanceof CodecReader codecReader) {
            KnnVectorsReader knnVectorsReader = codecReader.getVectorReader();
            if (knnVectorsReader instanceof PerFieldKnnVectorsFormat.FieldsReader perFieldVectorsReader) {
                knnVectorsReader = perFieldVectorsReader.getFieldReader(field);
            }
            if (knnVectorsReader instanceof VectorsScorer scorer) {
                if (floatVector != null) {
                    return scorer.getRandomVectorScorer(field, floatVector);
                } else {
                    return scorer.getRandomVectorScorer(field, byteVector);
                }
            }
        }
        return null;
    }
```

So, I am gonna keep it as this approximate search for now, and we can optimize later as the interface is improved in Lucene.

Reply:
I see, thanks for explaining. We need the codec to own an internal representation of the query vector.